File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(bf.readLine());
        schedule = new int[n][2];
        dp = new int[n];
        for (int i = 0; i < dp.length; i++) {
            dp[i] = -1;
        }
        for (int i = 0; i < n; i++) {
            String[] split = bf.readLine().split(" ");
            schedule[i][0] = Integer.parseInt(split[0]);
            schedule[i][1] = Integer.parseInt(split[1]);
        }
        System.out.println(recursive(0));
    }

    static int n;
    static int[][] schedule;
    static int[] dp;

    static private int recursive(int index) {
        if (index == n) {
            return 0;
        }
        if (index > n) {
            return Integer.MIN_VALUE;
        }
        if (dp[index] != -1) {
            return dp[index];
        }
        dp[index] = Math.max(
                recursive(index + schedule[index][0]) + schedule[index][1],
                recursive(index + 1)
        );

        return dp[index];
    }
}

File: 백준/Silver/14501. 퇴사/퇴사.py
from sys import stdin
input = stdin.readline

def solution(schedule:list):
    n = len(schedule)
    dp = [0] * (n+1)

    for i in range(n-1, -1, -1):
        if i + schedule[i][0] > n:
            dp[i] = dp[i+1]
        else:    
            dp[i] += max(dp[i+1], schedule[i][1]+dp[i+schedule[i][0]])

    return dp[0]

N = int(input())
schedule = [list(map(int, input().split())) for _ in range(N)]

print(solution(schedule))